src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -111,58 +129,56 @@ removeIgnored fp = withIgnoreMode . cataNode remove
-        (Link, LINK {})        -> put ignoreNone $> defNode
-        (Link, _)            ->
-          case getIgnoreMode node of
-            Just mode'  -> do
-              lift . tell $ makeError modePos fp mode ""
-              handleMode node mode'
-            Nothing     -> Node pos ty <$> sequence subs
-
+      get >>= \case
         -- When no `Ignore` state is set check next node for annotation,
         -- if found then set it as new `IgnoreMode` otherwise skip node.
-        (None, _)              ->
-          case getIgnoreMode node of
-            Just mode' -> handleMode node mode'
-            Nothing    -> Node pos ty <$> sequence subs
-
-    handleMode
-      :: Node
-      -> IgnoreMode
+        Nothing                    -> handleIgnoreMode pos ty subs $ getIgnoreMode node
+        Just (Ignore mode modePos) ->
+          case (mode, ty) of
+            -- We expect to find a paragraph immediately after the
+            -- `ignore paragraph` annotanion. If the paragraph is not
+            -- found we should report an error.
+            (Paragraph, PARAGRAPH) -> put Nothing $> defNode
+            (Paragraph, x)         -> do
+              lift . tell . makeError modePos fp . ParagraphErr $ prettyType x
+              put Nothing
+              Node pos ty <$> sequence subs
+
+            -- We don't expect to find an `ignore file` annotation here,
+            -- since that annotation should be at the top of the file and
+            -- the file should already be ignored when `checkIgnoreFile` is called.
+            -- We should report an error if we find it anyway.
+            (File, _)              -> do
+              lift . tell $ makeError modePos fp FileErr
+              put Nothing
+              Node pos ty <$> sequence subs
+
+            -- When we find an `ignore link` annotation, we skip nodes until
+            -- we find a link and ignore it, or we find another ignore annotation,
+            -- then we should report an error and set new `Ignore` state.
+            (Link, LINK {})        -> put Nothing $> defNode
+            (Link, _)              -> do
+              let ignoreMode = getIgnoreMode node
+              unless (ignoreMode == NotAnAnnotation) $
+                lift . tell $ makeError modePos fp LinkErr
+              handleIgnoreMode pos ty subs ignoreMode
+
+    handleIgnoreMode
+      :: Maybe PosInfo
+      -> NodeType
+      -> [ScannerM Node]
+      -> GetIgnoreMode
       -> ScannerM Node
-    handleMode node = \case
-      -- Report unknown `IgnoreMode`.
-      None   -> do
-        let unrecognised = fromMaybe ""
-              $ safeHead . drop 1 . words =<< getXrefcheckContent node
-        lift . tell $ makeError (getPosition node) fp None unrecognised
-        put ignoreNone $> defNode
-      -- Set new `Ignore` state.
-      mode'  -> put (Ignore mode' $ getPosition node) $> defNode
+    handleIgnoreMode pos mode subs = \case
```

Comment:
The second arg is named `mode` but its type is `NodeType` - let's rename the arg to `nodeType` (and then the `mode'` variable in line 174 can be renamed to just `mode`)